consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,436 @@
+	Broadcast(msg M)
+
+	// SendMsg would send a message to a specific validator. This would be required for helping send resquest and
+	// response message to help a specifc validator to catch up.
+	SendMsg(validatorAddr A, msg M)
+}
+
+type Listeners[V Hashable[H], H Hash, A Addr] struct {
+	ProposalListener  Listener[Proposal[V, H, A], V, H, A]
+	PrevoteListener   Listener[Prevote[H, A], V, H, A]
+	PrecommitListener Listener[Precommit[H, A], V, H, A]
+}
+
+type Broadcasters[V Hashable[H], H Hash, A Addr] struct {
+	ProposalBroadcaster  Broadcaster[Proposal[V, H, A], V, H, A]
+	PrevoteBroadcaster   Broadcaster[Prevote[H, A], V, H, A]
+	PrecommitBroadcaster Broadcaster[Precommit[H, A], V, H, A]
+}
+
+type Tendermint[V Hashable[H], H Hash, A Addr] struct {
+	nodeAddr A
+
+	state state[V, H] // Todo: Does state need to be protected?
+
+	messages       messages[V, H, A]
+	futureMessages messages[V, H, A]
+
+	futureMessagesMu *sync.Mutex
+
+	timeoutPropose   timeoutFn
+	timeoutPrevote   timeoutFn
+	timeoutPrecommit timeoutFn
+
+	application Application[V, H]
+	blockchain  Blockchain[V, H, A]
+	validators  Validators[A]
+
+	listeners    Listeners[V, H, A]
+	broadcasters Broadcasters[V, H, A]
+
+	scheduledTms []timeout
+	timeoutsCh   chan timeout
+
+	// Future round messages are sent to the loop through the following channels
+	proposalsCh  chan Proposal[V, H, A]
+	prevotesCh   chan Prevote[H, A]
+	precommitsCh chan Precommit[H, A]
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+type state[V Hashable[H], H Hash] struct {
+	height uint
+	round  uint
+	step   step
+
+	lockedValue *V
+	lockedRound int
+	validValue  *V
+	validRound  int
+
+	// The following are round level variable therefore when a round changes they must be reset.
+	line34Executed bool
+	line36Executed bool
+	line47Executed bool
+}
+
+func New[V Hashable[H], H Hash, A Addr](addr A, app Application[V, H], chain Blockchain[V, H, A], vals Validators[A],
+	listeners Listeners[V, H, A], broadcasters Broadcasters[V, H, A], tmPropose, tmPrevote, tmPrecommit timeoutFn,
+) *Tendermint[V, H, A] {
+	return &Tendermint[V, H, A]{
+		nodeAddr: addr,
+		state: state[V, H]{
+			height:      chain.Height(),
+			lockedRound: -1,
+			validRound:  -1,
+		},
+		messages:         newMessages[V, H, A](),
+		futureMessages:   newMessages[V, H, A](),
+		futureMessagesMu: &sync.Mutex{},
+		timeoutPropose:   tmPropose,
+		timeoutPrevote:   tmPrevote,
+		timeoutPrecommit: tmPrecommit,
+		application:      app,
+		blockchain:       chain,
+		validators:       vals,
+		listeners:        listeners,
+		broadcasters:     broadcasters,
+		scheduledTms:     make([]timeout, 0),
+		timeoutsCh:       make(chan timeout),
+		proposalsCh:      make(chan Proposal[V, H, A]),
+		prevotesCh:       make(chan Prevote[H, A]),
+		precommitsCh:     make(chan Precommit[H, A]),
+		quit:             make(chan struct{}),
+	}
+}
+
+func (t *Tendermint[V, H, A]) Start() {
+	t.wg.Add(1)
```

Comment:
I'm wondering if we can remove wait group here. By doing so, we can let the caller decides how to run the algorithm.

Reply:
I don't think I follow.

The `Start()` is not blocking, I add to the wait group so that when the user calls `Stop()` then the go routine that is started in the `Start()` returns before `Stop()` returns.